File: backend/engines/php/match-all.php
#!php

<?php

function exitWithResponse($response) {
    echo json_encode($response, JSON_PRETTY_PRINT | JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    exit(isset($response['error']) ? 1 : 0);
}

$engineValue = @$argv[1];
$text = @$argv[2];
$patternValue = @$argv[3];
$flagsValue = @$argv[4];

if (!is_string($patternValue)) {
    exitWithResponse(['engineValue' => $engineValue, 'error' => ['message' => 'invalid pattern']]);
}

if (!is_string($text)) {
    exitWithResponse(['engineValue' => $engineValue, 'error' => ['message' => 'invalid text']]);
}

// $pattern = strpos($patternValue, '/') === false ? "/$patternValue/$flagsValue" : $patternValue;
$pattern = "/$patternValue/$flagsValue";

$result = preg_match_all($pattern, $text, $results, PREG_OFFSET_CAPTURE);

if (!$result) {
  exitWithResponse(['engineValue' => $engineValue, 'matches' => []]);
}

$matches = array_map(function ($result) {
  return [
    'index' => $result[1],
    'substring' => $result[0],
  ];
}, $results[0]);

exitWithResponse(['engineValue' => $engineValue, 'matches' => $matches]);

File: backend/engines/ruby/match-all.rb
#!ruby -W0

require 'json'

engineValue = ARGV[0].nil? ? '' : ARGV[0]
text = ARGV[1]
patternValue = ARGV[2]
flagsValue = ARGV[3].nil? ? '' : ARGV[3]

def exitWithResponse(response)
    puts JSON.pretty_generate(response)
    exit response.key?(:error) ? 1 : 0
end

if patternValue.nil?
    exitWithResponse({engineValue: engineValue, error: {message: 'invalid pattern'}})
end

if text.nil?
    exitWithResponse({engineValue: engineValue, error: {message: 'invalid text'}})
end

# patternValue =~ /^\/(.*)\/(.*)$/m
# pattern = '(' + ($1.nil? ? patternValue : $1) + ')'
# flags = $2.nil? ? '' : $2

begin
    regexpObject = flagsValue == '' ? Regexp.new('(' + patternValue + ')') : Regexp.new('((?' + flagsValue + ')' + patternValue + ')')
rescue
    exitWithResponse({engineValue: engineValue, error: {message: 'invalid pattern'}})
end

matches = text.enum_for(:scan, regexpObject).map {
  |result| {index: Regexp.last_match.begin(0), substring: result.kind_of?(Array) ? result[0] : result}
}
exitWithResponse({engineValue: engineValue, matches: matches})
